### ISSUE TITLE
Support ES6 iterables?

### ISSUE THREAD

**Issue by lukehorvat** [petkaantonov/bluebird#569]
Have you considered supporting [ES6 iteration protocols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols)?

Would be cool to do something like this:

```javascript
}).then(() => {
  return new Set([1, 2, 3, 4]).values(); // Returns an iterable, not an array.
}).each(value => {
  console.log(value);
});
```

At the moment, that code throws an `expecting an array, a promise or a thenable` error, so you have to do `Array.from(new Set([1, 2, 3, 4]).values())` instead.

**Comment by benjamingr** [comment]
We considered this, but we decided to defer the decision to when someone would actually ask for it (Petka didn't want to add the feature before it was asked for or used).

I think this is a good candidate for 3.0 (with a fast array path). What do you think @petkaantonov ?

**Comment by petkaantonov** [comment]
Chrome dev supports them so they'll be on stable by end of june, I guess that's a good deadline.

**Comment by domenic** [comment]
Pretty sure Chrome stable already supports them.






On Wed, Apr 8, 2015 at 12:58 AM -0700, "Petka Antonov" <notifications@github.com<mailto:notifications@github.com>> wrote:


Chrome dev supports them so they'll be on stable by end of june, I guess that's a good deadline.

-

**Comment by benjamingr** [comment]
@domenic 

```js
> Promise.all(new Set([1, 2, 3, 4]).values())

```

> Uncaught (in promise) TypeError: invalid_argument {stack: (...), message: "invalid_argument"}message: "invalid_argument"

**Comment by petkaantonov** [comment]
@domenic Current stable is 41, which uses 4.1 but promise.all and race supporting iterables is implemented in V8 4.3.x

**Comment by domenic** [comment]
Right. I meant iterables in general, not the fix to the Promise statics.

**Comment by spion** [comment]
+1 on this, it will enable using `Promise.each` for iteration w/o the overhead of making new arrays

```es6
Promise.each(range(1,n), (k) => ...)
```

**Comment by za-creature** [comment]
@spion could I get some feedback on my [non-es6 proposal](https://github.com/petkaantonov/bluebird/issues/734) to your problem?

**Comment by petkaantonov** [comment]
Fixed in 3.0 release
